dan_layer/indexer_lib/src/transaction_autofiller.rs

```diff
@@ -0,0 +1,110 @@
+//   Copyright 2023 The Tari Project
+//   SPDX-License-Identifier: BSD-3-Clause
+
+use std::{fmt::Display, sync::Arc};
+
+use tari_dan_common_types::ShardId;
+use tari_engine_types::substate::{Substate, SubstateAddress};
+use tari_transaction::{SubstateChange, Transaction};
+use tari_validator_node_rpc::client::{SubstateResult, ValidatorNodeClientFactory};
+
+use crate::{
+    committee_provider::CommitteeProvider,
+    error::IndexerError,
+    substate_decoder::{find_related_substates, SubstateDecoderError},
+    substate_scanner::SubstateScanner,
+};
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransactionAutofillerError {
+    #[error("Could not decode the substate: {0}")]
+    SubstateDecoderError(#[from] SubstateDecoderError),
+    #[error("Indexer error: {0}")]
+    IndexerError(#[from] IndexerError),
+}
+
+pub struct TransactionAutofiller<TCommitteeProvider, TVnClient> {
+    substate_scanner: Arc<SubstateScanner<TCommitteeProvider, TVnClient>>,
+}
+
+impl<TCommitteeProvider, TVnClient> TransactionAutofiller<TCommitteeProvider, TVnClient>
+where
+    TCommitteeProvider: CommitteeProvider,
+    TVnClient: ValidatorNodeClientFactory<Addr = TCommitteeProvider::Addr>,
+    TCommitteeProvider::Addr: Display,
+{
+    pub fn new(substate_scanner: Arc<SubstateScanner<TCommitteeProvider, TVnClient>>) -> Self {
+        Self { substate_scanner }
+    }
+
+    pub async fn autofill_transaction(
+        &self,
+        original_transaction: &Transaction,
+    ) -> Result<Transaction, TransactionAutofillerError> {
+        // we will include the inputs and outputs into the "involved_objects" field
+        // note that the transaction hash will not change as the "involved_objects" is not part of the hash
+        let mut autofilled_transaction = original_transaction.clone();
+
+        // scan the network to fetch all the substates for each required input
+        // TODO: perform this loop concurrently by spawning a tokio task for each scan
+        let mut input_addresses: Vec<(SubstateAddress, u32)> = vec![];
+        let mut input_substates: Vec<Substate> = vec![];
+        for r in autofilled_transaction.meta().required_inputs() {
+            let scan_res = self.substate_scanner.get_substate(r.address(), r.version()).await?;
+
+            // TODO: should we return an error if some of the inputs are not "Up"?
```

Comment:
Maybe a warning? It might be useful to allow this method to not fail during testing, or if we pass a flag `allow_downed_inputs`

Reply:
Added a warning and also refactored the input substate fetching (to differentiate if a specific version was required). I think we can skip returning errors and just log a warning, as if the input is actually needed it will not be included anyway and consensus will fail. If in the future we see that we want to fail sooner we can add more complex logic with a parameter here.